From 03a8ec2d4ebd883ad2a1bea7c07ca16cf85491c7 Mon Sep 17 00:00:00 2001
From: Ewan Milne <emilne@redhat.com>
Date: Wed, 20 Aug 2014 18:58:45 -0400
Subject: [scsi] sd: Reduce buffer size for vpd request

Message-id: <1408561176-18133-7-git-send-email-emilne@redhat.com>
Patchwork-id: 88423
O-Subject: [RHEL7.1 PATCH e-stor 06/57] [SCSI] sd: Reduce buffer size for vpd request
Bugzilla: 1132092
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: "Ewan D. Milne" <emilne@redhat.com>

upstream commit af73623f5f10eb3832c87a169b28f7df040a875b

Tested by Ewan D. Milne and Barry Donahue with a combination
of manual and automated testing.  No issues seen.

upstream commit af73623f5f10eb3832c87a169b28f7df040a875b
Author: Bernd Schubert <bernd.schubert@itwm.fraunhofer.de>
Date:   Mon Sep 23 14:47:32 2013 +0200

    [SCSI] sd: Reduce buffer size for vpd request

    Somehow older areca firmware versions have issues with
    scsi_get_vpd_page() and a large buffer, the firmware
    seems to crash and the scsi error-handler will start endless
    recovery retries.
    Limiting the buf-size to 64-bytes fixes this issue with older
    firmware versions (<1.49 for my controller).

    Fixes a regression with areca controllers and older firmware versions
    introduced by commit: 66c28f97120e8a621afd5aa7a31c4b85c547d33d

    Reported-by: Nix <nix@esperi.org.uk>
    Tested-by: Nix <nix@esperi.org.uk>
    Signed-off-by: Bernd Schubert <bernd.schubert@itwm.fraunhofer.de>
    Cc: stable@vger.kernel.org # delay inclusion for 2 months for testing
    Acked-by: Martin K. Petersen <martin.petersen@oracle.com>
    Signed-off-by: James Bottomley <JBottomley@Parallels.com>

Signed-off-by: Ewan D. Milne <emilne@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/sd.c b/drivers/scsi/sd.c
index 917113c..35bc29e 100644
--- a/drivers/scsi/sd.c
+++ b/drivers/scsi/sd.c
@@ -2633,13 +2633,16 @@ static void sd_read_write_same(struct scsi_disk *sdkp, unsigned char *buffer)
  }
 
  if (scsi_report_opcode(sdev, buffer, SD_BUF_SIZE, INQUIRY) < 0) {
+  /* too large values might cause issues with arcmsr */
+  int vpd_buf_len = 64;
+
   sdev->no_report_opcodes = 1;
 
   /* Disable WRITE SAME if REPORT SUPPORTED OPERATION
    * CODES is unsupported and the device has an ATA
    * Information VPD page (SAT).
    */
-  if (!scsi_get_vpd_page(sdev, 0x89, buffer, SD_BUF_SIZE))
+  if (!scsi_get_vpd_page(sdev, 0x89, buffer, vpd_buf_len))
    sdev->no_write_same = 1;
  }
 
-- 
1.7.1